File: mano_memory/ManoMemory.py
from m5.params import *
from m5.SimObject import SimObject

class ManoMemory(SimObject):
    type = 'ManoMemory'
    cxx_header = "CAProject/mano_memory/mano_memory.hh"
    cxx_class = "gem5::ManoMemory"

    ar = Param.AddressRegister("The Address Register connected to the memory")
    meminit = VectorParam.UInt16("Some data to initialize the memory")

File: mano_memory/mano_memory.cc
#include "CAProject/mano_memory/mano_memory.hh"
#include <array>

namespace gem5
{

ManoMemory::ManoMemory(const ManoMemoryParams &params) : 
    SimObject(params),
    ManoMemory_(),
    AR(*params.ar)
{
    std::vector<uint16_t> init_values = params.meminit;
    for (size_t i = 0; i < init_values.size() && i < ManoMemory_.size(); i++)
    {
        ManoMemory_[i] = init_values[i];
    }
}

void
ManoMemory::Write(uint16_t data)
{
    uint16_t address = AR.getValue();
    ManoMemory_[address] = data;
}

uint16_t
ManoMemory::Read()
{
    return ManoMemory_[AR.getValue()];
}

} // namespace gem5
